### ISSUE TITLE
New Direct Message endpoint needs to be implemented

### ISSUE THREAD

**Issue by melikyuksel** [bear/python-twitter#587]
Twitter deprecated the old direct message endpoints on September 17, 2018 (3 days ago) after warning about it in July. (see this link: https://twittercommunity.com/t/details-and-what-to-expect-from-the-api-deprecations-this-week-on-august-16-2018/110746)

Does Python Twitter yet support the new endpoints for it? (https://developer.twitter.com/en/docs/direct-messages/sending-and-receiving/guides/direct-message-migration)

Using the PostDirectMessage function right now results in "Sorry, that page does not exist." error 34 from Twitter.

**Comment by webtweakers** [comment]
I'm having the same issue, found the same cause, was looking for the same reason...

**Comment by at0m1sk** [comment]
https://github.com/bear/python-twitter/pull/579

There is a pull request that seems to integrate new endpoint. Request is from Aug 2018.

**Comment by spacecab07** [comment]
Any idea on when this might get resolved?

**Comment by jmingtan** [comment]
Yes the version in the pull request is working for us. There are breaking changes to the function though, which is forced by Twitter's changes in the new direct message endpoint. Probably the biggest change is that screen name is no longer supported as a parameter, so we need to call api.GetUser first to get the user id.

**Comment by ganmedia** [comment]
Any news about this? When will the pull request be integrated?

**Comment by bear** [comment]
i'll look at this now

**Comment by jeremylow** [comment]
added everything and fixed up the tests, changed the docs a few hours ago
to the 4.0 dev branch. sorry, been out of commission lately :(


Jeremy Low
JeremyLow@gmail.com

**Comment by bear** [comment]
@jeremylow do you want me to merge this to the master and bump the version to 3.6 since it's fixing a broken twitter?

**Comment by jeremylow** [comment]
I was thinking a major version bump, since we have to deprecate a bunch of
endpoints, but I can do 3.6 too. I'm not married to either option.


Jeremy Low
845.594.2392
JeremyLow@gmail.com

**Comment by bear** [comment]
@jeremylow I think I can make the method signature stay the same to reduce the impact - if `screen_name` is nil then make the call, if not use it

ok, i'll merge the PR and see if that fixes things

**Comment by bear** [comment]
I think I have it working and i've tagged a pre-release for v3.5

Can you all give this a go and see if it works - i'll hold off pushing this to PyPI until someone sanity checks it

**Comment by webtweakers** [comment]
Thanks for picking this up. I've checked the v3.5 release and now things work again, as expected! Looking good from my side.

**Comment by ganmedia** [comment]
Version 3.5 checked and everything working with no problems. 
Great job.

**Comment by kennethdamica** [comment]
PostDirectMessage works for me now, but I'm still getting error 34 for GetDirectMessages and GetSentDirectMessages. Looks like these endpoints have been collapsed into "GET direct_messages/events/list". Is this supposed to work in 3.5?

**Comment by spacecab07** [comment]
Sorry if this is a noob question. So, i need the user_id of the recipient of the DM?

**Comment by jeremylow** [comment]
@spacecab07 if you don’t supply it, the api will attempt to look up the user id from the screen name.

**Comment by spacecab07** [comment]
@jeremylow so should be ok to leave as 'user_id=None, screen_name=twitter_handle' ?

**Comment by webtweakers** [comment]
@spacecab07 Yes, I have already confirmed that this works well - also check the code change. The API needs to make an extra call to Twitter now, so it is probably wise to retrieve and store the user_id in your own application once and use that at every DM post.

**Comment by spacecab07** [comment]
@webtweakers ok, thanks .

**Comment by at0m1sk** [comment]
> 
> 
> PostDirectMessage works for me now, but I'm still getting error 34 for GetDirectMessages and GetSentDirectMessages. Looks like these endpoints have been collapsed into "GET direct_messages/events/list". Is this supposed to work in 3.5?

Same here, I am at 3.5 can't GET , error 34 for GetDM and GetSentDM

**Comment by MHlinux** [comment]
Have the same issue and still do not get it done 
api.PostDirectMessage(directmsg, user_id="xxx")
api.PostDirectMessage(directmsg, screen_name="xxx")

But all other works fine like 
api.GetFriends(screen_name="xxx")
api.GetFriends(user_id="xxx")

So what else do I have to do in addition for direct mesages ???

**Comment by LuccoJ** [comment]
Could this please be made a priority for pushing a release to Pypi? Currently, according to what I've been able to figure out, there is no Twitter wrapper in Pypi that actually supports direct messages. I assume there's a number of applications that are badly broken by this and not just mine.

**Comment by Artfulme** [comment]
Hi, backing up @LuccoJ on this one, you're just one signature way from releasing the working code to pypi.  Don't want to be pushy either, I have much respect for all the work you guys put in to maintain the code, but I have been following this thread since its inception and working direct messages can't come soon enough.

**Comment by jeremylow** [comment]
My apologies y'all; been a draining couple weeks. I thought the release had been uploaded to PyPI. I'll take care of it this weekend.

**Comment by jeremylow** [comment]
3.5 is now on pypi.

**Comment by Artfulme** [comment]
Thanks!

**Comment by webtweakers** [comment]
A little follow-up on this... It's possible that the Twitter API returns in error and that 'event' is not returned in the response.  Trying to create the DirectMessage as the response will then fail:

```python
dm = DirectMessage(
    created_at=data['event']['created_timestamp'],
    id=data['event']['id'],
    recipient_id=data['event']['message_create']['target']['recipient_id'],
    sender_id=data['event']['message_create']['sender_id'],
    text=data['event']['message_create']['message_data']['text'],
)
```

This code fails with a `KeyError: 'event'`.

This happens, for example, when attempting to send a direct message to a user that is not following you.

When setting return_json=True, the returned data will then be something like this:

```
{'errors': [{'code': 150,
   'message': 'You cannot send messages to users who are not following you.'}]}
```

Clearly 'event' is missing.

A fix would be to check if any errors have been returned, and/or check if 'event' actually exists in the returned data, before creating the DirectMessage object.

**Comment by pacu** [comment]
I'm using 3.5 and I'm still getting the error 34. I also tried to install from github with same results. Am I'm missing something? How can I help?

**Comment by JunksFiPenny** [comment]
Started getting this error message last week, and as a relative noob to how the Twitter module work, am having issues figuring out if the problem is same as above. No changes were made to code, no updates to module etc; simply stopped working...

![Capture](https://user-images.githubusercontent.com/32553838/63597355-43323280-c58b-11e9-8286-5a6da0dc2714.PNG)
